// @polkadot/dev-ts/cjs/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loaderOptions = exports.EXT_JSON_REGEX = exports.EXT_JS_REGEX = exports.EXT_TS_REGEX = exports.EXT_TS_ARRAY = exports.MOD_PATH = exports.CWD_URL = exports.CWD_PATH = void 0;
const tslib_1 = require("tslib");
const node_path_1 = tslib_1.__importDefault(require("node:path"));
const node_process_1 = tslib_1.__importDefault(require("node:process"));
const node_url_1 = require("node:url");
/** The path we are being executed from */
exports.CWD_PATH = node_process_1.default.cwd();
/** The cwd path we are being executed from in URL form */
exports.CWD_URL = (0, node_url_1.pathToFileURL)(`${exports.CWD_PATH}/`);
/** The root path to node_modules (assuming it is in the root) */
exports.MOD_PATH = node_path_1.default.join(exports.CWD_PATH, 'node_modules');
/** List of allowed extensions for mappings */
exports.EXT_TS_ARRAY = ['.ts', '.tsx'];
/** RegEx for files that we support via this loader */
exports.EXT_TS_REGEX = /\.tsx?$/;
/** RegEx for matching JS files (imports map to TS) */
exports.EXT_JS_REGEX = /\.jsx?$/;
/** RegEx for json files (as actually aliassed in polkadot-js) */
exports.EXT_JSON_REGEX = /\.json$/;
/** Options for loader config */
exports.loaderOptions = {};
