// @polkadot/dev/rootJs/Clazz.js
var _Clazz_something;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
class Clazz {
    /**
     * @param and the number we should and with
     */
    constructor(and) {
        _Clazz_something.set(this, 123456789);
        /**
         * @description Sets something to something
         * @param something The addition
         */
        this.setSomething = (something) => {
            __classPrivateFieldSet(this, _Clazz_something, (something ?? 123456) & this.and, "f");
            return __classPrivateFieldGet(this, _Clazz_something, "f");
        };
        this.and = and;
        __classPrivateFieldSet(this, _Clazz_something, __classPrivateFieldGet(this, _Clazz_something, "f") & and, "f");
    }
    get something() {
        return __classPrivateFieldGet(this, _Clazz_something, "f");
    }
    async doAsync() {
        const res = await new Promise((resolve) => resolve(true));
        console.log(res);
        return res;
    }
    toString() {
        return `something=${__classPrivateFieldGet(this, _Clazz_something, "f")}`;
    }
}
_Clazz_something = new WeakMap();
Clazz.staticProperty = 'foobar';
Clazz.staticFunction = () => Clazz.staticProperty;
export { Clazz };
