// @polkadot/dev/cjs/rootJs/Clazz.js
"use strict";
var _Clazz_something;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Clazz = void 0;
const tslib_1 = require("tslib");
class Clazz {
    /**
     * @param and the number we should and with
     */
    constructor(and) {
        _Clazz_something.set(this, 123456789);
        /**
         * @description Sets something to something
         * @param something The addition
         */
        this.setSomething = (something) => {
            tslib_1.__classPrivateFieldSet(this, _Clazz_something, (something ?? 123456) & this.and, "f");
            return tslib_1.__classPrivateFieldGet(this, _Clazz_something, "f");
        };
        this.and = and;
        tslib_1.__classPrivateFieldSet(this, _Clazz_something, tslib_1.__classPrivateFieldGet(this, _Clazz_something, "f") & and, "f");
    }
    get something() {
        return tslib_1.__classPrivateFieldGet(this, _Clazz_something, "f");
    }
    async doAsync() {
        const res = await new Promise((resolve) => resolve(true));
        console.log(res);
        return res;
    }
    toString() {
        return `something=${tslib_1.__classPrivateFieldGet(this, _Clazz_something, "f")}`;
    }
}
_Clazz_something = new WeakMap();
Clazz.staticProperty = 'foobar';
Clazz.staticFunction = () => Clazz.staticProperty;
exports.Clazz = Clazz;
